test/common/upstream/load_balancer_benchmark.cc

```diff
@@ -0,0 +1,137 @@
+  }
+
+  PrioritySetImpl priority_set_;
+  std::shared_ptr<MockClusterInfo> info_{new NiceMock<MockClusterInfo>()};
+  Stats::IsolatedStoreImpl stats_store_;
+  ClusterStats stats_{ClusterInfoImpl::generateStats(stats_store_)};
+  NiceMock<Runtime::MockLoader> runtime_;
+  Runtime::RandomGeneratorImpl random_;
+  Optional<envoy::api::v2::Cluster::RingHashLbConfig> config_;
+  std::unique_ptr<RingHashLoadBalancer> ring_hash_lb_;
+};
+
+static void BM_RingHashLoadBalancerBuildRing(benchmark::State& state) {
+  for (auto _ : state) {
+    state.PauseTiming();
+    const uint64_t num_hosts = state.range(0);
+    const uint64_t min_ring_size = state.range(1);
+    RingHashTester tester(num_hosts, min_ring_size);
+    state.ResumeTiming();
+
+    // We are only interested in timing the initial ring build.
+    tester.ring_hash_lb_->initialize();
+  }
+}
+BENCHMARK(BM_RingHashLoadBalancerBuildRing)
+    ->Args({100, 65536})
+    ->Args({200, 65536})
+    ->Args({500, 65536})
+    ->Args({100, 256000})
+    ->Args({200, 256000})
+    ->Args({500, 256000})
+    ->Unit(benchmark::kMillisecond);
+
+class TestLoadBalancerContext : public LoadBalancerContext {
+public:
+  // Upstream::LoadBalancerContext
+  Optional<uint64_t> computeHashKey() override { return hash_key_; }
+  const Router::MetadataMatchCriteria* metadataMatchCriteria() const override { return nullptr; }
+  const Network::Connection* downstreamConnection() const override { return nullptr; }
+
+  Optional<uint64_t> hash_key_;
+};
+
+static void BM_RingHashLoadBalancerChooseHost(benchmark::State& state) {
+  for (auto _ : state) {
+    // Do not time the creation of the ring.
+    state.PauseTiming();
+    const uint64_t num_hosts = state.range(0);
+    const uint64_t min_ring_size = state.range(1);
+    const uint64_t keys_to_simulate = state.range(2);
+    RingHashTester tester(num_hosts, min_ring_size);
+    tester.ring_hash_lb_->initialize();
+    LoadBalancerPtr lb = tester.ring_hash_lb_->factory()->create();
+    std::unordered_map<std::string, uint64_t> hit_counter;
+    TestLoadBalancerContext context;
+    state.ResumeTiming();
+
+    // Note: To a certain extent this is benchmarking the performance of xxhash as well as
+    // std::unordered_map. However, it should be roughly equivilant to the work done when
+    // comparing to different hashing algorithms such as Maglev.
+    for (uint64_t i = 0; i < keys_to_simulate; i++) {
+      context.hash_key_.value(
+          HashUtil::xxHash64(absl::string_view(reinterpret_cast<const char*>(&i), sizeof(i))));
+      hit_counter[lb->chooseHost(&context)->address()->asString()] += 1;
+    }
+
+    // Do not time computation of mean, standard deviation, and relative standard deviation.
+    state.PauseTiming();
+    double mean = 0;
+    std::for_each(hit_counter.begin(), hit_counter.end(),
+                  [&mean](const auto pair) { mean += pair.second; });
```

Comment:
In Envoy, is this preferred to
```
  for (const auto pair : hit_counter) {
     mean += pair.second;
  }
```

Reply:
I just copied this code from elsewhere. Will fix.